File: Python/merge_sort.py


def mergeSort(arr): 
	if len(arr) >1: 
		mid = len(arr)//2 
		t1 = arr[:mid] 
		t2 = arr[mid:] 

		mergeSort(t1) 
		mergeSort(t2) 

		i = j = k = 0
		
		while i < len(t1) and j < len(t2): 
			if t1[i] < t2[j]: 
				arr[k] = t1[i] 
				i+=1
			else: 
				arr[k] = t2[j] 
				j+=1
			k+=1

		while i < len(t1): 
			arr[k] = t1[i] 
			i+=1
			k+=1
		
		while j < len(t2): 
			arr[k] = t2[j] 
			j+=1
			k+=1



l = list(map(int,input().split()))

mergeSort(l) 

print(l)

File: C++/mergesort.cpp
#include <bits/stdc++.h> 
using namespace std; 

template<typename T>
void msfunc(vector<T> &array, int left, int right){
  if(left >= right) return;

  int mid = (left + right) / 2;
  msfunc(array, left, mid);
  msfunc(array, mid+1, right);

  int x = left, y = mid+1;
  while(x <= mid and y <= right){
    if(array[x] <= array[y]){
      x++;
    }else{
      int v = array[y];
      int i = y;

      for(; i != x; i--){
        array[i] = array[i-1];
      }

      array[x] = v;
      x++;
      mid++;
      y++;
    }
  }
}

template<typename T>
void mergesort(vector<T> &array){
  int n = array.size();
  msfunc(array, 0, n-1);
}

int main() 
{ 
  srand((unsigned int)time(NULL));
  vector<int> arr;
  for(int i = 0; i < 10; i++) arr.emplace_back(rand()%1000);
  mergesort(arr);
  for(int i = 0; i < arr.size(); i++){
    cout << arr[i] << " ";
  }
  cout << endl;
	return 0; 
} 
